File: _streamer/js/whip-client.js
/**
 * WHIPClient - WebRTC HTTP Ingest Protocol client
 * Handles streaming to WHIP-compatible servers
 * Optimized for single stream publishing
 */
export class WHIPClient {
    constructor(endpoint, updateStatusCallback) {
        if (!endpoint) {
            throw new Error('WHIP endpoint URL is required');
        }
        this.endpoint = endpoint;
        this.updateStatus = updateStatusCallback || ((msg) => console.log(msg));
        
        this.peerConnection = new RTCPeerConnection({
            iceServers: [
                { urls: 'stun:stun.l.google.com:19302' },
                {
                    urls: 'turn:159.65.180.44:3478',
                    username: 'voodooturn',
                    credential: 'vT_s7r0ngP@sswrd_4U'
                }
            ],
            iceCandidatePoolSize: 10,
            bundlePolicy: 'max-bundle',
            rtcpMuxPolicy: 'require',
            sdpSemantics: 'unified-plan'
        });

        // Store track senders for replacement
        this.videoSender = null;
        this.audioSender = null;

        // Add connection state handlers for better error reporting
        this.peerConnection.onconnectionstatechange = () => {
            const state = this.peerConnection.connectionState;
            this.updateStatus(`connection state: ${state}`);
        };

        this.peerConnection.oniceconnectionstatechange = () => {
            const state = this.peerConnection.iceConnectionState;
            this.updateStatus(`ICE connection state: ${state}`);

            if (state === 'failed') {
                this.restartIce();
            }
        };
    }

    async replaceTrack(track) {
        try {
            if (!this.peerConnection) return;

            // Handle null track (remove video)
            if (track === null) {
                if (this.videoSender) {
                    await this.videoSender.replaceTrack(null);
                    this.updateStatus(`video track removed - audio only mode`);
                }
                return;
            }

            let sender;
            if (track.kind === 'video') {
                if (!this.videoSender) {
                    this.videoSender = this.peerConnection.getSenders().find(s => s.track?.kind === 'video');
                }
                sender = this.videoSender;
            } else if (track.kind === 'audio') {
                if (!this.audioSender) {
                    this.audioSender = this.peerConnection.getSenders().find(s => s.track?.kind === 'audio');
                }
                sender = this.audioSender;
            }

            if (sender) {
                await sender.replaceTrack(track);

                // Update encoding parameters for video tracks
                if (track.kind === 'video') {
                    const params = sender.getParameters();
                    let videoBitrate, fps;
                    
                    // Different settings for HD vs SD
                    videoBitrate = parseInt(document.getElementById('videoBitrate').value);
                    fps = parseInt(document.getElementById('fps').value);

                    // Force bitrate settings
                    params.encodings = [{
                        maxBitrate: videoBitrate * 1000,
                        maxFramerate: fps,
                        scaleResolutionDownBy: 1.0,
                        active: true,
                        networkPriority: "high",
                        priority: "high",
                        degradationPreference: "maintain-framerate"
                    }];

                    await sender.setParameters(params);
                    
                    console.log(`video encoding parameters set:`, params.encodings[0]);
                }

                this.updateStatus(`${track.kind} track replaced successfully`);
            }
        } catch (error) {
            this.updateStatus(`Error replacing ${track.kind} track: ${error.message}`, true);
            console.error(`track replacement error:`, error);
        }
    }

    async restartIce() {
        try {
            await this.peerConnection.restartIce();
            const offer = await this.peerConnection.createOffer({ iceRestart: true });
            await this.peerConnection.setLocalDescription(offer);
        } catch (error) {
            this.updateStatus(`Failed to restart ICE: ${error.message}`, true);
        }
    }

    async startStream(stream) {
        try {
            console.log(`Starting stream with ${stream.getTracks().length} tracks`);
            
            // Store video sender reference for bitrate adjustments
            this.videoSender = null;
            
            // Add tracks to the peer connection with appropriate priority
            const senders = stream.getTracks().map(track => {
                const sender = this.peerConnection.addTrack(track, stream);

                // Log track settings for debugging
                if (track.kind === 'video') {
                    const settings = track.getSettings();
                    console.log(`🎬 WHIP: Adding video track with settings:`, {
                        width: settings.width,
                        height: settings.height,
                        frameRate: settings.frameRate,
                        aspectRatio: settings.aspectRatio,
                        deviceId: settings.deviceId
                    });
                }

                // Set selected codec and priority for video tracks
                if (track.kind === 'video') {
                    const transceivers = this.peerConnection.getTransceivers();
                    const transceiver = transceivers.find(t => t.sender === sender);
                    if (transceiver) {
                        const codecs = RTCRtpSender.getCapabilities('video').codecs;
                        const selectedCodec = document.getElementById('videoCodec').value;
                        
                        console.log(`- Looking for codec: video/${selectedCodec.toLowerCase()}`);
                        
                        const preferredCodecs = codecs.filter(codec => 
                            codec.mimeType.toLowerCase() === `video/${selectedCodec.toLowerCase()}`);
                        
                        if (preferredCodecs.length > 0) {
                            transceiver.setCodecPreferences([
                                ...preferredCodecs,
                                ...codecs.filter(c => !preferredCodecs.includes(c))
                            ]);
                            console.log(`✅ set codec preference to ${selectedCodec}`);
                        } else {
                            console.warn(`❌ ${selectedCodec} codec not supported by browser`);
                        }
                        transceiver.direction = 'sendonly';

                        // Enhanced video sender parameters
                        const params = sender.getParameters();
                        let videoBitrate, fps;
                        
                        // Different settings for HD vs SD
                        videoBitrate = parseInt(document.getElementById('videoBitrate').value);
                        fps = parseInt(document.getElementById('fps').value);

                        params.encodings = [{
                            maxBitrate: videoBitrate * 1000,
                            maxFramerate: fps,
                            scaleResolutionDownBy: 1.0,
                            active: true,
                            networkPriority: "high",
                            priority: "high",
                            degradationPreference: "maintain-framerate"
                        }];

                        sender.setParameters(params);
                        console.log(`initial video encoding parameters:`, params.encodings[0]);
                        
                        // Store video sender for later use
                        this.videoSender = sender;
                    }
                }
                return sender;
            });

            // Create and set local description with specific constraints
            const offer = await this.peerConnection.createOffer({
                offerToReceiveAudio: false,
                offerToReceiveVideo: false,
                voiceActivityDetection: false,
                iceRestart: true
            });

            // Modify SDP to set bitrates and priorities
            offer.sdp = this.modifySDPWithBitrate(offer.sdp);

            await this.peerConnection.setLocalDescription(offer);
            await this.waitForIceGathering();

            const response = await fetch(this.endpoint, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/sdp',
                },
                body: this.peerConnection.localDescription.sdp
            });

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const answer = await response.text();
            await this.peerConnection.setRemoteDescription({
                type: 'answer',
                sdp: answer
            });

            // Log the negotiated configuration and enforce bitrate
            const videoSender = this.peerConnection.getSenders().find(s => s.track?.kind === 'video');
            if (videoSender) {
                const params = await videoSender.getParameters();
                console.log(`negotiated video parameters:`, params.encodings);
                
                // Enforce bitrate again after negotiation
                this.enforceBitrate();
                
                // Set up periodic bitrate enforcement
                this.bitrateInterval = setInterval(() => {
                    this.enforceBitrate();
                }, 2000); // Every 2 seconds
            }

            console.log(`stream started successfully`);

        } catch (error) {
            console.error(`streaming failed:`, error);
            throw error;
        }
    }
    
    async enforceBitrate() {
        if (!this.videoSender || !this.peerConnection || this.peerConnection.connectionState !== 'connected') {
            return;
        }
        
        try {
            let targetBitrate;
            targetBitrate = parseInt(document.getElementById('videoBitrate').value);
            
            const params = this.videoSender.getParameters();
            
            if (params.encodings && params.encodings[0]) {
                // Check current bitrate
                const currentMaxBitrate = params.encodings[0].maxBitrate;
                const targetBitrateKbps = targetBitrate * 1000;
                
                // Only update if different
                if (currentMaxBitrate !== targetBitrateKbps) {
                    params.encodings[0].maxBitrate = targetBitrateKbps;
                    await this.videoSender.setParameters(params);
                    console.log(`enforced bitrate: ${targetBitrate} kbps`);
                }
            }
        } catch (error) {
            console.error(`Error enforcing bitrate:`, error);
        }
    }

    modifySDPWithBitrate(sdp) {
        const lines = sdp.split('\r\n');
        const newLines = [];
        let isVideoSection = false;
        let isAudioSection = false;
        
        // Different bitrate settings for HD vs SD
        let videoBitrate, fps;
        videoBitrate = parseInt(document.getElementById('videoBitrate').value);
        fps = parseInt(document.getElementById('fps').value);

        for (let line of lines) {
            // Detect video section
            if (line.startsWith('m=video')) {
                isVideoSection = true;
                isAudioSection = false;
            } else if (line.startsWith('m=audio')) {
                isVideoSection = false;
                isAudioSection = true;
            } else if (line.startsWith('m=')) {
                isVideoSection = false;
                isAudioSection = false;
            }

            // Change sendrecv to sendonly
            if (line.includes('a=sendrecv')) {
                newLines.push('a=sendonly');
                continue;
            }

            // Add the current line
            newLines.push(line);

            // Add video-specific settings
            if (isVideoSection && line.startsWith('c=')) {
                // Bandwidth settings
                newLines.push(`b=AS:${videoBitrate}`);
                newLines.push(`b=TIAS:${videoBitrate * 1000}`);

                // Transport feedback
                newLines.push('a=rtcp-fb:* transport-cc');
                newLines.push('a=rtcp-fb:* nack');
                newLines.push('a=rtcp-fb:* nack pli');
                newLines.push('a=rtcp-fb:* ccm fir');
                newLines.push('a=rtcp-fb:* goog-remb');

                // Priority settings based on stream type
                newLines.push('a=content:main');
                newLines.push('a=priority:high');
                newLines.push('a=x-google-max-bitrate:' + (videoBitrate * 1000));
                newLines.push('a=x-google-min-bitrate:' + (videoBitrate * 1000));
                newLines.push('a=x-google-start-bitrate:' + (videoBitrate * 1000));
                newLines.push('a=x-google-cpu-overuse-detection:false');

                // Additional performance settings
                newLines.push('a=x-google-buffer-latency:0');
                newLines.push('a=x-google-enable-pre-encode-drop:false');

                // Codec specific settings
                const selectedCodec = document.getElementById('videoCodec').value;
                
                // Find the actual payload type for the codec
                let codecPayloadType = null;
                for (let i = 0; i < newLines.length; i++) {
                    if (selectedCodec === 'VP9' && newLines[i].includes('VP9/90000')) {
                        const match = newLines[i].match(/^a=rtpmap:(\d+) VP9\/90000/);
                        if (match) codecPayloadType = match[1];
                        break;
                    } else if (selectedCodec === 'H264' && newLines[i].includes('H264/90000')) {
                        const match = newLines[i].match(/^a=rtpmap:(\d+) H264\/90000/);
                        if (match) codecPayloadType = match[1];
                        break;
                    } else if (selectedCodec === 'VP8' && newLines[i].includes('VP8/90000')) {
                        const match = newLines[i].match(/^a=rtpmap:(\d+) VP8\/90000/);
                        if (match) codecPayloadType = match[1];
                        break;
                    } else if (selectedCodec === 'AV1' && (newLines[i].includes('AV01/90000') || newLines[i].includes('av01/90000'))) {
                        const match = newLines[i].match(/^a=rtpmap:(\d+) (AV01|av01)\/90000/i);
                        if (match) codecPayloadType = match[1];
                        break;
                    }
                }
                
                if (codecPayloadType) {
                    if (selectedCodec === 'VP9') {
                        newLines.push(`a=fmtp:${codecPayloadType} profile-id=0`);
                        newLines.push(`a=fmtp:${codecPayloadType} max-fs=12288`);
                        newLines.push(`a=fmtp:${codecPayloadType} max-fr=${fps}`);
                    } else if (selectedCodec === 'H264') {
                        newLines.push(`a=fmtp:${codecPayloadType} level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f`);
                    } else if (selectedCodec === 'VP8') {
                        newLines.push(`a=fmtp:${codecPayloadType} max-fs=12288`);
                        newLines.push(`a=fmtp:${codecPayloadType} max-fr=${fps}`);
                    } else if (selectedCodec === 'AV1') {
                        newLines.push(`a=fmtp:${codecPayloadType} profile=0`);
                        newLines.push(`a=fmtp:${codecPayloadType} level=5.1`);
                        newLines.push(`a=fmtp:${codecPayloadType} tier=0`);
                    }
                }
            }
        }

        // Handle audio section for Studio Sound
        return this.modifySDPForStudioSound(newLines);
    }

    modifySDPForStudioSound(lines) {
        const studioSoundEnabled = document.getElementById('studioSound')?.checked;
        let opusPayloadType = null;

        // Find the payload type for Opus (typically opus/48000/2)
        for (let i = 0; i < lines.length; i++) {
            if (lines[i].includes('opus/48000/2')) {
                const match = lines[i].match(/^a=rtpmap:(\d+) opus\/48000\/2/);
                if (match && match[1]) {
                    opusPayloadType = match[1];
                    console.log(`Found Opus payload type: ${opusPayloadType}`);
                    break;
                }
            }
        }

        // Modify the fmtp line for Opus based on studioSound setting
        if (opusPayloadType) {
            for (let i = 0; i < lines.length; i++) {
                if (lines[i].startsWith(`a=fmtp:${opusPayloadType}`)) {
                    console.log(`Original Opus fmtp: ${lines[i]}`);
                    // Remove existing stereo parameters first to avoid duplicates
                    let fmtpLine = lines[i].replace(/;\s*stereo=1/g, '').replace(/;\s*sprop-stereo=1/g, '');

                    if (studioSoundEnabled) {
                        // Add stereo parameters if Studio Sound is ON
                        if (!fmtpLine.includes('usedtx=0')) { // dtx=1 is incompatible with stereo=1
                             fmtpLine += ';stereo=1;sprop-stereo=1';
                             console.log(`Added stereo parameters for Studio Sound.`);
                        } else {
                             console.warn(`Cannot enable stereo with DTX enabled. Keeping mono.`);
                        }
                    } else {
                        console.log(`Ensuring mono parameters (Studio Sound OFF).`);
                    }

                    lines[i] = fmtpLine;
                    console.log(`Modified Opus fmtp: ${lines[i]}`);
                    break;
                }
            }
        } else {
            console.warn(`Could not find Opus payload type.`);
        }

        return lines.join('\r\n');
    }

    async waitForIceGathering() {
        return new Promise((resolve) => {
            if (this.peerConnection.iceGatheringState === 'complete') {
                resolve();
            } else {
                const checkState = () => {
                    if (this.peerConnection.iceGatheringState === 'complete') {
                        this.peerConnection.removeEventListener('icegatheringstatechange', checkState);
                        resolve();
                    }
                };
                this.peerConnection.addEventListener('icegatheringstatechange', checkState);
                setTimeout(resolve, 3000);
            }
        });
    }

    stopStream() {
        // Clear bitrate enforcement interval
        if (this.bitrateInterval) {
            clearInterval(this.bitrateInterval);
            this.bitrateInterval = null;
        }
        
        if (this.peerConnection) {
            this.peerConnection.close();
            this.peerConnection = null;
        }
        
        this.videoSender = null;
        this.audioSender = null;
        
        console.log(`stream stopped`);
    }
}

File: voodoovideo/WHIPClient.swift
import WebRTC
import Foundation

/**
 * WHIPClient - WebRTC HTTP Ingest Protocol client for Swift
 * Direct port of whip-client.js functionality using WebRTC framework
 */
class WHIPClient: NSObject {
    
    // MARK: - Properties
    private let endpoint: String
    private var peerConnection: RTCPeerConnection?
    private var videoSender: RTCRtpSender?
    private var audioSender: RTCRtpSender?
    private var bitrateEnforcementTimer: Timer?
    private let updateStatusCallback: (String, Bool) -> Void
    
    // WebRTC configuration optimized for WHIP streaming
    private let rtcConfiguration: RTCConfiguration = {
        let config = RTCConfiguration()
        config.iceServers = [
            RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"]),
            RTCIceServer(urlStrings: ["stun:stun1.l.google.com:19302"]),
            RTCIceServer(urlStrings: ["stun:stun2.l.google.com:19302"]),
            RTCIceServer(
                urlStrings: ["turn:159.65.180.44:3478"],
                username: "voodooturn",
                credential: "vT_s7r0ngP@sswrd_4U"
            )
        ]
        config.iceCandidatePoolSize = 0  // Disable candidate pooling for WHIP
        config.bundlePolicy = .maxCompatible  // Most compatible option
        config.rtcpMuxPolicy = .require
        config.sdpSemantics = .unifiedPlan
        config.continualGatheringPolicy = .gatherOnce  // Gather ICE candidates once
        return config
    }()
    
    // Settings for encoding
    var videoBitrate: Int = 3000 // kbps
    var frameRate: Int = 30
    var videoCodec: String = "H264" // H264, VP9, VP8, AV1
    
    // MARK: - Initialization
    
    init(endpoint: String, updateStatusCallback: @escaping (String, Bool) -> Void) {
        guard !endpoint.isEmpty else {
            fatalError("WHIP endpoint URL is required")
        }
        
        self.endpoint = endpoint
        self.updateStatusCallback = updateStatusCallback
        super.init()
        
        setupPeerConnection()
    }
    
    // MARK: - Private Setup Methods
    
    private func setupPeerConnection() {
        let factory = RTCPeerConnectionFactory()
        peerConnection = factory.peerConnection(with: rtcConfiguration, constraints: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil), delegate: self)
        
        updateStatusCallback("WHIP client initialized", false)
    }
    
    // MARK: - Public Methods
    
    func startStream(videoTrack: RTCVideoTrack?, audioTrack: RTCAudioTrack?) async throws {
        guard let peerConnection = peerConnection else {
            throw WHIPError.invalidState("Peer connection not initialized")
        }
        
        updateStatusCallback("Starting WHIP stream...", false)
        
        // Add tracks to the peer connection
        if let videoTrack = videoTrack {
            videoSender = peerConnection.add(videoTrack, streamIds: ["stream"])
            updateStatusCallback("Added video track", false)
            
            // Configure video encoding
            try await configureVideoSender()
        }
        
        if let audioTrack = audioTrack {
            audioSender = peerConnection.add(audioTrack, streamIds: ["stream"])
            updateStatusCallback("Added audio track", false)
        }
        
        // Create offer with simpler constraints
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: nil,
            optionalConstraints: [
                "OfferToReceiveAudio": "false",
                "OfferToReceiveVideo": "false"
            ]
        )
        let offer = try await peerConnection.offer(for: constraints)
        
        // Modify SDP for WHIP requirements
        var modifiedSDP = modifySDPWithBitrate(offer.sdp)
        modifiedSDP = modifySDPForSendOnly(modifiedSDP)
        
        let modifiedOffer = RTCSessionDescription(type: .offer, sdp: modifiedSDP)
        try await peerConnection.setLocalDescription(modifiedOffer)
        
        // Wait for ICE gathering
        try await waitForIceGathering()
        
        // Send WHIP request
        let answer = try await sendWHIPRequest(peerConnection.localDescription!.sdp)
        
        // Set remote description
        try await peerConnection.setRemoteDescription(RTCSessionDescription(type: .answer, sdp: answer))
        
        // Start bitrate enforcement
        startBitrateEnforcement()
        
        updateStatusCallback("WHIP stream connected", false)
    }
    
    func stopStream() {
        updateStatusCallback("Stopping WHIP stream...", false)
        
        // Stop bitrate enforcement
        bitrateEnforcementTimer?.invalidate()
        bitrateEnforcementTimer = nil
        
        // Close peer connection
        peerConnection?.close()
        peerConnection = nil
        
        // Reset senders
        videoSender = nil
        audioSender = nil
        
        updateStatusCallback("WHIP stream stopped", false)
    }
    
    func replaceVideoTrack(_ track: RTCVideoTrack?) async throws {
        guard let peerConnection = peerConnection else {
            throw WHIPError.invalidState("No peer connection available")
        }
        // Remove existing video sender if present
        if let sender = videoSender {
            peerConnection.removeTrack(sender)
            videoSender = nil
        }
        // Add new video track if provided
        if let newTrack = track {
            videoSender = peerConnection.add(newTrack, streamIds: ["stream"])
            updateStatusCallback("Video track replaced", false)
            try await configureVideoSender()
        } else {
            updateStatusCallback("Video track removed", false)
        }
    }
    
    func replaceAudioTrack(_ track: RTCAudioTrack?) async throws {
        guard let peerConnection = peerConnection else {
            throw WHIPError.invalidState("No peer connection available")
        }
        // Remove existing audio sender if present
        if let sender = audioSender {
            peerConnection.removeTrack(sender)
            audioSender = nil
        }
        // Add new audio track if provided
        if let newTrack = track {
            audioSender = peerConnection.add(newTrack, streamIds: ["stream"])
            updateStatusCallback("Audio track replaced", false)
        } else {
            updateStatusCallback("Audio track removed", false)
        }
    }
    
    // MARK: - Private Methods
    
    private func configureVideoSender() async throws {
        guard let videoSender = videoSender else { return }
        
        let parameters = videoSender.parameters
        
        // Configure encoding parameters
        if !parameters.encodings.isEmpty {
            parameters.encodings[0].maxBitrateBps = NSNumber(value: videoBitrate * 1000)
            parameters.encodings[0].maxFramerate = NSNumber(value: frameRate)
            parameters.encodings[0].scaleResolutionDownBy = NSNumber(value: 1.0)
            parameters.encodings[0].isActive = true
            parameters.encodings[0].networkPriority = .high
        }
        
        // Setting parameters is no longer supported directly on the sender.
        // The encoding parameters should be configured before adding the sender or may not be configurable at all.
        print("⚠️ WHIPClient: Directly setting sender parameters is not supported in this version. Configure encoding parameters before adding the sender.")
        
        updateStatusCallback("Video encoding configured: \(videoBitrate)kbps @ \(frameRate)fps", false)
    }
    
    private func startBitrateEnforcement() {
        bitrateEnforcementTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
            Task {
                await self?.enforceBitrate()
            }
        }
    }
    
    private func enforceBitrate() async {
        guard let videoSender = videoSender,
              let peerConnection = peerConnection,
              peerConnection.connectionState == .connected else {
            return
        }
        
        do {
            let parameters = videoSender.parameters
            if !parameters.encodings.isEmpty {
                let currentBitrate = parameters.encodings[0].maxBitrateBps?.intValue ?? 0
                let targetBitrate = videoBitrate * 1000
                
                if currentBitrate != targetBitrate {
                    // Directly setting parameters is no longer supported.
                    // Dynamic bitrate enforcement via parameter adjustment is not available.
                    print("⚠️ WHIPClient: Dynamic bitrate enforcement is not supported. Skipping parameter update.")
                }
            }
        } catch {
            print("❌ WHIPClient: Error enforcing bitrate: \(error)")
        }
    }
    
    private func modifySDPWithBitrate(_ sdp: String) -> String {
        var lines = sdp.components(separatedBy: "\r\n")
        var newLines: [String] = []
        var isVideoSection = false
        
        for line in lines {
            if line.hasPrefix("m=video") {
                isVideoSection = true
            } else if line.hasPrefix("m=") && !line.hasPrefix("m=video") {
                isVideoSection = false
            }
            
            newLines.append(line)
            
            // Add video-specific settings
            if isVideoSection && line.hasPrefix("c=") {
                // Add bandwidth settings
                newLines.append("b=AS:\(videoBitrate)")
                newLines.append("b=TIAS:\(videoBitrate * 1000)")
                
                // Add transport feedback
                newLines.append("a=rtcp-fb:* transport-cc")
                newLines.append("a=rtcp-fb:* nack")
                newLines.append("a=rtcp-fb:* nack pli")
                newLines.append("a=rtcp-fb:* ccm fir")
                newLines.append("a=rtcp-fb:* goog-remb")
                
                // Priority settings
                newLines.append("a=content:main")
                newLines.append("a=priority:high")
                newLines.append("a=x-google-max-bitrate:\(videoBitrate * 1000)")
                newLines.append("a=x-google-min-bitrate:\(videoBitrate * 1000)")
                newLines.append("a=x-google-start-bitrate:\(videoBitrate * 1000)")
            }
        }
        
        return newLines.joined(separator: "\r\n")
    }
    
    private func modifySDPForSendOnly(_ sdp: String) -> String {
        return sdp.replacingOccurrences(of: "a=sendrecv", with: "a=sendonly")
    }
    
    private func waitForIceGathering() async throws {
        guard let peerConnection = peerConnection else {
            throw WHIPError.invalidState("Peer connection not available")
        }
        
        if peerConnection.iceGatheringState == .complete {
            return
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            var observer: NSObjectProtocol?
            
            let checkState = {
                if peerConnection.iceGatheringState == .complete {
                    if let observer = observer {
                        NotificationCenter.default.removeObserver(observer)
                    }
                    continuation.resume()
                }
            }
            
            observer = NotificationCenter.default.addObserver(
                forName: NSNotification.Name("RTCPeerConnectionIceGatheringStateChanged"),
                object: peerConnection,
                queue: .main
            ) { _ in
                checkState()
            }
            
            // Timeout after 5 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 5.0) {
                if let observer = observer {
                    NotificationCenter.default.removeObserver(observer)
                }
                continuation.resume()
            }
            
            // Check immediately in case already complete
            checkState()
        }
    }
    
    private func sendWHIPRequest(_ sdp: String) async throws -> String {
        guard let url = URL(string: endpoint) else {
            throw WHIPError.invalidEndpoint("Invalid WHIP endpoint URL")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/sdp", forHTTPHeaderField: "Content-Type")
        request.httpBody = sdp.data(using: .utf8)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw WHIPError.networkError("Invalid response type")
        }
        
        guard httpResponse.statusCode == 200 || httpResponse.statusCode == 201 else {
            throw WHIPError.httpError("HTTP \(httpResponse.statusCode)")
        }
        
        guard let answerSDP = String(data: data, encoding: .utf8) else {
            throw WHIPError.networkError("Invalid SDP response")
        }
        
        return answerSDP
    }
}

// MARK: - RTCPeerConnectionDelegate

extension WHIPClient: RTCPeerConnectionDelegate {
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
        updateStatusCallback("Signaling state: \(stateChanged)", false)
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        // Not used for WHIP (send-only)
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
        // Not used for WHIP (send-only)
    }
    
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
        // Handle renegotiation if needed
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        let stateDescription: String
        switch newState {
        case .new:
            stateDescription = "new"
        case .checking:
            stateDescription = "checking"
        case .connected:
            stateDescription = "connected"
        case .completed:
            stateDescription = "completed"
        case .failed:
            stateDescription = "failed"
        case .disconnected:
            stateDescription = "disconnected"
        case .closed:
            stateDescription = "closed"
        case .count:
            stateDescription = "count"
        @unknown default:
            stateDescription = "unknown"
        }
        
        print("🔗 WHIPClient: ICE connection state changed to: \(stateDescription)")
        updateStatusCallback("ICE: \(stateDescription)", newState == .failed)
        
        if newState == .connected || newState == .completed {
            updateStatusCallback("WHIP stream connected", false)
        } else if newState == .failed {
            updateStatusCallback("ICE connection failed - check network/TURN server", true)
            print("❌ WHIPClient: ICE connection failed. Check:")
            print("   - Network connectivity")
            print("   - TURN server credentials")
            print("   - Firewall settings")
        } else if newState == .disconnected {
            updateStatusCallback("ICE connection lost", true)
        }
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {
        updateStatusCallback("ICE gathering state: \(newState)", false)
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        // ICE candidates are handled automatically
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {
        // Handle candidate removal if needed
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        // Not used for WHIP
    }
}

// MARK: - Error Types

enum WHIPError: LocalizedError {
    case invalidState(String)
    case invalidEndpoint(String)
    case networkError(String)
    case httpError(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidState(let message):
            return "Invalid state: \(message)"
        case .invalidEndpoint(let message):
            return "Invalid endpoint: \(message)"
        case .networkError(let message):
            return "Network error: \(message)"
        case .httpError(let message):
            return "HTTP error: \(message)"
        }
    }
}

